zetaclient/chains/solana/observer/outbound.go

```diff
@@ -2,22 +2,344 @@ package observer
+	if txResult.Meta.Err != nil {
+		logger.Error().Msg("tx is not successful")
+		return nil, false
+	}
+
+	// parse gateway instruction from tx result
+	inst, err := ob.ParseGatewayInstruction(txResult, coinType)
+	if err != nil {
+		logger.Error().Err(err).Msg("ParseGatewayInstruction err")
+		return nil, false
+	}
+	txNonce := inst.GatewayNonce()
+
+	// recover ECDSA signer from instruction
+	signerECDSA, err := inst.Signer()
+	if err != nil {
+		logger.Error().Err(err).Msg("cannot get instruction signer")
+		return nil, false
+	}
+
+	// check tx authorization
+	if signerECDSA != ob.TSS().EVMAddress() {
+		logger.Error().Msgf("tx signer %s is not matching TSS address %s", signerECDSA, ob.TSS().EVMAddress())
+		return nil, false
+	}
+
+	// check tx nonce
+	if txNonce != nonce {
+		logger.Error().Msgf("tx nonce %d is not matching cctx nonce %d", txNonce, nonce)
+		return nil, false
+	}
+
+	return txResult, true
+}
+
+// ParseGatewayInstruction parses the instruction signer and nonce from tx result
+func (ob *Observer) ParseGatewayInstruction(
+	txResult *rpc.GetTransactionResult,
+	coinType coin.CoinType,
+) (contract.OutboundInstruction, error) {
+	// unmarshal transaction
+	tx, err := txResult.Transaction.GetTransaction()
+	if err != nil {
+		return nil, errors.Wrap(err, "error unmarshaling transaction")
+	}
+
+	// there should be only one single instruction ('withdraw' or 'withdraw_spl_token')
+	if len(tx.Message.Instructions) != 1 {
+		return nil, fmt.Errorf("want 1 instruction, got %d", len(tx.Message.Instructions))
+	}
+	instruction := tx.Message.Instructions[0]
+
+	// get the program ID
+	programPk, err := tx.Message.Program(instruction.ProgramIDIndex)
+	if err != nil {
+		return nil, errors.Wrap(err, "error getting program ID")
+	}
+
+	// the instruction should be an invocation of the gateway program
+	if !programPk.Equals(ob.gatewayID) {
+		return nil, errors.New("not a gateway program invocation")
+	}
+
+	// parse the instruction as a 'withdraw' or 'withdraw_spl_token'
+	switch coinType {
+	case coin.CoinType_Gas:
+		return ob.ParseInstructionWithdraw(tx, 0)
+	default:
+		return nil, errors.New("unsupported outbound coin type")
+	}
+}
+
+// ParseInstructionWithdraw tries to parse an instruction as a 'withdraw'.
+// It returns nil if the instruction can't be parsed as a 'withdraw'.
+func (ob *Observer) ParseInstructionWithdraw(
+	tx *solana.Transaction,
+	instructionIndex int,
+) (*contract.WithdrawInstructionParams, error) {
+	// locate instruction by index
+	instruction := tx.Message.Instructions[instructionIndex]
+
+	// try deserializing instruction as a 'withdraw'
+	inst := &contract.WithdrawInstructionParams{}
+	err := borsh.Deserialize(inst, instruction.Data)
+	if err != nil {
+		return nil, errors.Wrap(err, "error deserializing instruction")
+	}
+
+	// check the discriminator to ensure it's a 'withdraw' instruction
+	if inst.Discriminator != contract.DiscriminatorWithdraw() {
+		return nil, errors.New("not a withdraw instruction")
```

Comment:
Can we show values in error

Reply:
https://github.com/zeta-chain/node/pull/2560/commits/ada365437e0782d6d2f6725ca87d6c53a687118d